src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1124,30 +1143,21 @@ public bool TakeMessageHeaders(SocketInput input, FrameRequestHeaders requestHea
                     }
 
                     scan = beginValue;
-                    if (scan.Seek(ref _vectorCRs) == -1)
+                    if (scan.Seek(ref _vectorCRs, ref end) == -1)
                     {
-                        // no "\r" in sight, burn used bytes and go back to await more data
-                        return false;
+                        RejectRequest(RequestRejectionReason.MissingCRInHeaderLine);
                     }
 
                     scan.Take(); // we know this is '\r'
                     ch = scan.Take(); // expecting '\n'
 
-                    if (ch == -1)
-                    {
-                        return false;
-                    }
-                    else if (ch != '\n')
+                    if (ch != '\n')
                     {
-                        RejectRequest(RequestRejectionReason.HeaderLineMustEndInCRLFOnlyCRFound);
+                        RejectRequest(RequestRejectionReason.HeaderLineMustEndInCRLF);
                     }
 
                     var next = scan.Peek();
-                    if (next == -1)
-                    {
-                        return false;
```

Comment:
I don't think we reject linear whitespace if the input stalls right beforehand. Probably should add a test.